src/reducers/vms.js

```diff
@@ -77,11 +79,24 @@ const vms = actionReducer(initialState, {
     if (page) {
       st = st.set('page', page)
     }
+
+    const vmsIds = Object.keys(updates)
+    const filteredSet = new Set()
+    const missedVms = st.get('missedVms')
+    for (const vmId of vmsIds) {
+      if (!missedVms.has(vmId)) {
+        filteredSet.add(vmId)
+      }
+    }
+
+    st = st.set('missedVms', filteredSet)
+
     return st
   },
   [REMOVE_VMS] (state, { payload: { vmIds } }) {
     const mutable = state.asMutable()
     vmIds.forEach(vmId => mutable.deleteIn([ 'vms', vmId ]))
+    mutable.set('missedVms', new Set([ ...state.get('missedVms'), ...vmIds ]))
```

Comment:
Much easier if you utilize ImmutableJS APIs:
```suggestion
    mutable.set('missedVms', state.get('missedVms').union(vmIds)))
```